1.MLP_Binary_Classifier/MLP_utils/utils.py

```diff
@@ -0,0 +1,627 @@
+    )
+    optimizer = eval(optimizer)
+
+    early_stopping_patience = 15
+    early_stopping_counter = 0
+
+    train_acc = []
+    train_loss = []
+
+    valid_acc = []
+    valid_loss = []
+
+    total_step = len(train_loader)
+    total_step_val = len(valid_loader)
+
+    valid_loss_min = np.inf
+
+    epochs_ran = []
+
+    for epoch in range(EPOCHS):
+
+        epochs_ran.append(epoch + 1)
+
+        (
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            correct,
+            total,
+            correct_v,
+            total_v,
+        ) = train_n_validate(
+            model,
+            optimizer,
+            criterion,
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            total_step,
+            total_step_val,
+            params,
+            train_loader,
+            valid_loader,
+        )
+
+        if np.mean(valid_loss) <= valid_loss_min:
+            torch.save(model.state_dict(), f"./{date.today()}_state_dict.pt")
+            print(
+                f"Epoch {epoch + 0:01}: Validation loss decreased ({valid_loss_min:.6f} --> {np.mean(valid_loss):.6f}).  Saving model ..."
+            )
+            valid_loss_min = np.mean(valid_loss)
+            early_stopping_counter = 0  # reset counter if validation loss decreases
+        else:
+            print(f"Epoch {epoch + 0:01}: Validation loss did not decrease")
+            early_stopping_counter += 1
+
+        if early_stopping_counter > early_stopping_patience:
+            print("Early stopped at epoch :", epoch)
+            break
+
+        print(
+            f"\t Train_Loss: {np.mean(train_loss):.4f} Train_Acc: {(100 * correct / total):.3f} Val_Loss: {np.mean(valid_loss):.4f}  BEST VAL Loss: {valid_loss_min:.4f}  Val_Acc: {(100 * correct_v / total_v):.3f}\n"
+        )
+    return train_loss, train_acc, valid_loss, valid_acc, epochs_ran, model
+
+
+# Plot training data over epochs
+def plot_metric_vs_epoch(df, x, y1, y2):
+
+    sns.lineplot(x=x, y=y1, data=df)
+    sns.lineplot(x=x, y=y2, data=df)
+
+
+def test_optimized_model(model, test_loader, in_features, parameter_dict, params):
+    """
+    This function tests the trained optimized model on test dataset
+
+    Parameters:
+        None
+    Return:
+        y_pred_list : list
+            lsit of predicted values
+        y_pred_prob_list : list
+            lsit of probabilities for predicted values
+    """
+
+    model = optimized_model(in_features, parameter_dict)
+    model = model.to(params.DEVICE)
+    criterion = nn.BCEWithLogitsLoss()
+    optim_method = parameter_dict["optimizer"].strip("'")
+    optimizer = (
+        f'optim.{optim_method}(model.parameters(), lr=parameter_dict["learning_rate"])'
+    )
+    optimizer = eval(optimizer)
+
+    y_pred_prob_list = []
+    y_pred_list = []
+
+    # Loading the best model
+    # model.load_state_dict(torch.load('./state_dict.pt'))
```

Comment:
is this vestigial? If so, please remove. If not, consider adding a function argument to load state dictionary (or not)